File: Darts/src/code/screen/game/GamePanelGolf.java
package code.screen.game;

import code.ai.AbstractDartsModel;
import code.db.DartEntity;
import code.db.ParticipantEntity;
import code.object.Dart;
import object.HandyArrayList;
import object.HashMapList;
import util.DateUtil;

public class GamePanelGolf extends DartsGamePanel<DartsScorerGolf>
{
	//Number of rounds - 9 holes or 18?
	private int numberOfRounds = -1;
	
	public GamePanelGolf(DartsGameScreen parent)
	{
		super(parent);
	}
	
	@Override
	public void initImpl(String gameParams)
	{
		//The params tell us how many holes
		numberOfRounds = Integer.parseInt(gameParams);
	}
	
	@Override
	public void doAiTurn(AbstractDartsModel model)
	{
		int targetHole = currentRound.getRoundNumber();
		int dartNo = dartsThrown.size() + 1;
		model.throwGolfDart(targetHole, dartNo, dartboard);
	}

	@Override
	public void loadDartsForParticipant(int playerNumber, HashMapList<Integer, Dart> hmRoundToDarts, int lastRound)
	{
		DartsScorerGolf scorer = hmPlayerNumberToDartsScorer.get(playerNumber);
		for (int i=1; i<=lastRound; i++)
		{
			HandyArrayList<Dart> darts = hmRoundToDarts.get(i);
			scorer.addDarts(darts);
		}
	}

	@Override
	public void updateVariablesForNewRound(){}
	@Override
	public void resetRoundVariables(){}
	@Override
	public void updateVariablesForDartThrown(Dart dart){}

	@Override
	public boolean shouldStopAfterDartThrown()
	{
		if (dartsThrown.size() == 3)
		{
			return true;
		}
		
		int score = getScoreForMostRecentDart();
		if (activeScorer.getHuman())
		{
			return score == 1;
		}
		else
		{
			int noDarts = dartsThrown.size();
			
			AbstractDartsModel model = getCurrentPlayerStrategy();
			int stopThreshold = model.getStopThresholdForDartNo(noDarts);

			return score <= stopThreshold;
		}
	}
	private int getScoreForMostRecentDart()
	{
		Dart lastDart = dartsThrown.lastElement();
		
		int targetHole = currentRound.getRoundNumber();
		return lastDart.getGolfScore(targetHole);
	}

	@Override
	public void saveDartsAndProceed()
	{
		long roundId = currentRound.getRowId();
		for (int i=0; i<dartsThrown.size(); i++)
		{
			Dart dart = dartsThrown.get(i);
			DartEntity.factoryAndSave(dart, roundId, i+1, -1);
		}
		
		activeScorer.finaliseRoundScore();
		
		if (currentRound.getRoundNumber() == numberOfRounds)
		{
			handlePlayerFinish();
		}
		
		currentPlayerNumber = getNextPlayerNumber(currentPlayerNumber);
		if (getActiveCount() > 0)
		{
			nextTurn();
		}
		else
		{
			finishGame();
		}
	}
	
	@Override
	protected int getFinishingPositionFromPlayersRemaining()
	{
		//Finishing positions are determined at the end
		return -1;
	}
	private void finishGame()
	{
		//Get the participants sorted by score so we can assign finishing positions
		HandyArrayList<ParticipantEntity> participants = hmPlayerNumberToParticipant.getValuesAsVector();
		participants.sort((ParticipantEntity p1, ParticipantEntity p2) -> Integer.compare(p1.getFinalScore(), p2.getFinalScore()));
		
		int previousScore = Integer.MAX_VALUE;
		int finishPos = 1;
		for (int i=0; i<participants.size(); i++)
		{
			ParticipantEntity pt = participants.get(i);
			int ptScore = pt.getFinalScore();
			if (ptScore > previousScore)
			{
				finishPos++;
			}
			
			pt.setFinishingPosition(finishPos);
			pt.saveToDatabase();
			
			previousScore = pt.getFinalScore();
		}
		
		updateScorersWithFinishingPositions();
		
		gameEntity.setDtFinish(DateUtil.getSqlDateNow());
		gameEntity.saveToDatabase();
		
		parentWindow.startNextGameIfNecessary();
		
		allPlayersFinished();
	}

	@Override
	public DartsScorerGolf factoryScorer()
	{
		return new DartsScorerGolf();
	}

	@Override
	public boolean shouldAIStop()
	{
		return false;
	}
	
	@Override
	protected void doMissAnimation()
	{
		dartboard.doGolfMiss();
	}
}

File: Darts/src/burlton/dartzee/code/screen/game/GamePanelGolf.kt
package burlton.dartzee.code.screen.game

import burlton.core.code.obj.HashMapList
import burlton.dartzee.code.`object`.Dart
import burlton.dartzee.code.achievements.ACHIEVEMENT_REF_GOLF_COURSE_MASTER
import burlton.dartzee.code.achievements.ACHIEVEMENT_REF_GOLF_POINTS_RISKED
import burlton.dartzee.code.achievements.retrieveAchievementForDetail
import burlton.dartzee.code.ai.AbstractDartsModel
import burlton.dartzee.code.db.AchievementEntity
import burlton.dartzee.code.db.GameEntity
import burlton.dartzee.code.screen.Dartboard
import burlton.dartzee.code.screen.game.scorer.DartsScorerGolf

open class GamePanelGolf(parent: AbstractDartsGameScreen, game: GameEntity) : GamePanelFixedLength<DartsScorerGolf, Dartboard>(parent, game)
{
    //Number of rounds - 9 holes or 18?
    override val totalRounds = Integer.parseInt(game.gameParams)

    override fun factoryDartboard() = Dartboard()

    private fun getScoreForMostRecentDart() : Int
    {
        val lastDart = dartsThrown.last()

        val targetHole = currentRoundNumber
        return lastDart.getGolfScore(targetHole)
    }

    override fun initImpl(game: GameEntity) {}

    override fun doAiTurn(model: AbstractDartsModel)
    {
        val targetHole = currentRoundNumber
        val dartNo = dartsThrown.size + 1
        model.throwGolfDart(targetHole, dartNo, dartboard)
    }

    override fun loadDartsForParticipant(playerNumber: Int, hmRoundToDarts: HashMapList<Int, Dart>, totalRounds: Int)
    {
        val scorer = hmPlayerNumberToDartsScorer[playerNumber]
        for (i in 1..totalRounds)
        {
            val darts = hmRoundToDarts[i]!!
            scorer?.addDarts(darts)
        }
    }

    override fun updateVariablesForNewRound() {}
    override fun resetRoundVariables() {}
    override fun updateVariablesForDartThrown(dart: Dart) {}

    override fun shouldStopAfterDartThrown(): Boolean
    {
        if (dartsThrown.size == 3)
        {
            return true
        }

        val score = getScoreForMostRecentDart()
        if (activeScorer.human)
        {
            return score == 1
        }
        else
        {
            val noDarts = dartsThrown.size

            val model = getCurrentPlayerStrategy()
            val stopThreshold = model!!.getStopThresholdForDartNo(noDarts)

            return score <= stopThreshold
        }
    }

    override fun saveDartsAndProceed()
    {
        saveDartsToDatabase()

        activeScorer.finaliseRoundScore()

        unlockAchievements()

        finishRound()
    }

    fun unlockAchievements()
    {
        val dartsRisked = dartsThrown - dartsThrown.last()
        val pointsRisked = dartsRisked.map{ 5 - it.getGolfScore(currentRoundNumber) }.sum()

        if (pointsRisked > 0)
        {
            AchievementEntity.incrementAchievement(ACHIEVEMENT_REF_GOLF_POINTS_RISKED, getCurrentPlayerId(), gameEntity.rowId, pointsRisked)
        }

        val lastDart = dartsThrown.last()
        if (lastDart.getGolfScore(currentRoundNumber) == 1
         && retrieveAchievementForDetail(ACHIEVEMENT_REF_GOLF_COURSE_MASTER, getCurrentPlayerId(), "$currentRoundNumber") == null)
        {
            AchievementEntity.insertAchievement(ACHIEVEMENT_REF_GOLF_COURSE_MASTER, getCurrentPlayerId(), getGameId(), "$currentRoundNumber")
        }
    }


    override fun getFinishingPositionFromPlayersRemaining(): Int
    {
        //Finishing positions are determined at the end
        return -1
    }

    override fun factoryScorer() = DartsScorerGolf()

    override fun shouldAIStop() = false

    override fun doMissAnimation()
    {
        dartboard.doGolfMiss()
    }

    override fun factoryStatsPanel(): GameStatisticsPanel
    {
        return GameStatisticsPanelGolf()
    }
}
